engine/gamesys/src/gamesys/scripts/script_resource.cpp

```diff
@@ -2657,6 +2681,99 @@ static int SetSound(lua_State* L) {
     return 0;
 }
 
+static uint8_t* CheckBufferOrString(lua_State* L, int index, uint32_t* data_size)
+{
+    dmScript::LuaHBuffer* lua_buffer = dmScript::ToBuffer(L, index);
+    if (lua_buffer)
+    {
+        dmBuffer::HBuffer buffer = dmGameSystem::UnpackLuaBuffer(lua_buffer);
+
+        uint8_t* data = 0;
+        dmBuffer::GetBytes(buffer, (void**)&data, data_size);
+        return data;
+    }
+    else if (lua_isstring(L, index))
+    {
+        size_t string_len;
+        uint8_t* data = (uint8_t*)luaL_checklstring(L, -1, &string_len);
+        *data_size = (uint32_t)string_len;
+        return data;
+    }
+    luaL_error(L, "The field must be using either a buffer or a string.");
+    return 0;
+}
+
+static int CreateSoundData(lua_State* L)
+{
+    DM_LUA_STACK_CHECK(L, 1);
+
+    const char* path = luaL_checkstring(L, 1);
+    const char* path_ext = dmResource::GetExtFromPath(path);
+
+    const char* exts[]  = {".wav", ".ogg", ".wavc", ".oggc"};
+
+    dmhash_t canonical_path_hash = 0;
+    PreCreateResources(L, path, exts, DM_ARRAY_SIZE(exts), &canonical_path_hash);
+
+    // Find the correct type, as the resource types are registered as .wavc / .oggc
+    char type_suffix[32];
+    {
+        // Skip the '.'
+        int len = dmStrlCpy(type_suffix, path_ext+1, sizeof(type_suffix));
+        if (type_suffix[len-1] != 'c')
+        {
+            type_suffix[len++] = 'c';
+            type_suffix[len] = 0;
+        }
+    }
+
+    dmResource::HResourceType resource_type;
+    dmResource::Result r = dmResource::GetTypeFromExtension(g_ResourceModule.m_Factory, type_suffix, &resource_type);
```

Comment:
For http provider, we need to make sure the path is usable.
I.e. a general server will serve `.ogg`/`.wav` and not our own `.oggc` etc.

Hence the special code to preserve the path, yet still get the corresponding registered type.